### DISCUSSION TITLE
Sleep Tracking

### DISCUSSION THREAD

**Discussion by melvinredondotanis** [InfiniTimeOrg/InfiniTime#1918] [Ideas]
Do you think it could be possible to add this functionality according to the following points ? :
- start recording between certain times defined in the app (supported by gatebridge it seems)
- make a score updated at regular intervals
score depends on :
low heart rate + weak movements = deep sleep
down
high heart rate + high movement = light sleep

example formula:
bpm * (movement/2)

**Comment by thiswillbeyourgithub** [comment]
Shameless plug: https://github.com/thiswillbeyourgithub/sleep_tracker_pinetime_wasp-os

**Reply by melvinredondotanis** [reply]
Ok but it's for wasp-os, adding this functionality in infinitime would be more practical in my opinion.
I am a new user of Pinetime, I don't know the technical details of infinitime well but it should be possible to add this function?, I have some basic knowledge of C if that helps.

**Comment by Schweini07** [comment]
I am wondering, I found nothing on Github about this topic before this post, has this idea never come up before?
It is definitely something that should be implemented on InfiniTime, maybe someone can provide some further information on the current state of this?
Meanwhile I'll try to work out how to implement such an app, I was actually thinking about doing such a project before I got my PineTime :smile:

**Reply by melvinredondotanis** [reply]
It seems to me that the subject has already been broached, but to no avail.
For my part, I'm going to look for information on the role of heart rate and movement on sleep. 🕵️

**Reply by Schweini07** [reply]
That would be fantastic! I am already in the middle of developing a small prototype of this functionality, but of course it would be good to know how to best measure all this stuff :sweat_smile: .
What is also very interesting I think is the implementation of this functionality with companion apps, GadgetBridge for example allows for this. But this is a bit further down the line, right now I think a functioning prototype is enough :smiley:

**Reply by Schweini07** [reply]
Ah, I actually found an issue about this: https://github.com/InfiniTimeOrg/InfiniTime/issues/307
With an interesting Wikipedia article about it: https://en.wikipedia.org/wiki/Actigraphy

**Reply by melvinredondotanis** [reply]
indeed, they were concerned about the reliability of sensor support. Fortunately there has been progress although I find that it is still a little unstable

**Comment by tausen** [comment]
Just adding my 2 cents: 

There's a nice paper on the subject by Olivia Walch, specifically on the topic of sleep tracking with wearables using "raw" sensor data, intended to get some sleep tracking algorithms out in the free open source world:

https://academic.oup.com/sleep/article/42/12/zsz180/5549536

Quick summary: they subjected a bunch of people to polysomnography (the gold standard for sleep tracking, in her words) while wearing an apple watch, which they managed to pull raw accelerometer and heart rate data from. They then run a bunch of different classifiers on the data collected with the watch, comparing their performance using the data from the polysomnogram as the "ground truth". They put a lot of work into comparing how well the classifiers work depending on how many data sources they use (accelerometer, heart rate, circadian clock estimate). 

As far as I recall, the accelerometer data holds the most information and actually goes a long way by itself. Heart rate alone is not worth much, but combining the two is a good improvement over accelerometer alone. There's a slight improvement in also including an estimate of the users' circadian clock, for instance through an activity sensor such as the step counter.

I hacked together [a small Python script](https://gist.github.com/tausen/d79ebc7f62845f0159d0291533886155) using [bleak](https://github.com/hbldh/bleak) to collect data from my PineTime. I had to buy an external bluetooth adapter and a long usb extension cable to get a reliable enough connection between my computer and watch throughout the night, my laptop internal adapter was simply too poor. It then worked without a hitch, though. The logging here is on the extreme side, since I'm getting every single update on both the accelerometer and the heart rate:

![raw](https://github.com/InfiniTimeOrg/InfiniTime/assets/1283666/62297037-6b17-48ba-b22b-146840b08763)

Walch has been kind enough to make both the data they collected and her code publicly available:

https://github.com/ojwalch/sleep_classifiers

I tried plotting a few of her datasets, and they do resemble the plot above. I think the PineTime heart rate could use a bit of filtering, though (or perhaps it's just because my sample rate is a bit high). 

Her code is mainly focused on generating the figures for the paper, and I had to spend some time getting the right versions of dependencies and such. My idea was to dig out the actual preprocessing and the parameters of their model after being trained, then try repeating that processing on my data and running the classifier. Of course I don't have any ground truth to compare with, but I think I could borrow a Garmin watch to compare with if nothing else :)

A couple more papers I noted as candidates for future reading:
https://www.mdpi.com/1424-8220/21/5/1562
https://pubmed.ncbi.nlm.nih.gov/32215550/
https://ieeexplore.ieee.org/document/8879875
https://vbn.aau.dk/files/489314266/1_s2.0_S1389945722011327_main.pdf

**Reply by melvinredondotanis** [reply]
very interesting, I will read this article in more detail.
it's really cool, the test you did is very interesting. if I'm not mistaken you recorded your 8 hours of sleep. and pinetime does indeed seem technically capable of tracking sleep.

the different phases of sleep are visible:
1- low activity at the beginning
2- phases of deep sleep and paradoxical sleep:
3- and then waking up at the end of a beacon of paradoxical sleep

also I wouldn't have thought about it but actually the number of steps is a good element for following the circadian rhythm

**Reply by FintasticMan** [reply]
Which version of InfiniTime did you use for these tests? 1.13 has got a much better heart rate processing algorithm than previous versions, so it'd be interesting to see if that changes things much. The changes do mean that it rejects more values if the connection with the skin isn't good, which might happen during sleep.

**Reply by tausen** [reply]
This was tested with 1.13.0 + https://github.com/InfiniTimeOrg/InfiniTime/pull/1718 so I could get continuous heart rate updates.

**Reply by FintasticMan** [reply]
I have a theory that the biggest outliers could be the values just after it's started measuring. I.E. the measurements after it reports a heart rate of 0.

**Reply by tausen** [reply]
It's not unlikely there could be some poor skin contact throughout the test, I did not put much effort into ensuring a tight fit etc.

Perhaps another test with focus on heart rate measurements over a longer duration, not necessarily related to sleep tracking?

**Reply by melvinredondotanis** [reply]
I will try your script and try a few things to find the best settings. I will let you know in the coming days

**Reply by tausen** [reply]
Sounds good! If you need it, here's the bluetooth adapter I had success with: https://www.amazon.com/Bluetooth-Adapter-Wireless-Transmitter-Receiver/dp/B09KG7QQ5V

**Reply by melvinredondotanis** [reply]
thank you, but my pc already has bluetooth.
but how did you make the watch continue to send these measurements when the screen goes to sleep. are you using custom firmware?

**Reply by Schweini07** [reply]
This PR implements that feature but is not merged yet: https://github.com/InfiniTimeOrg/InfiniTime/pull/1718

**Reply by tausen** [reply]
> thank you, but my pc already has bluetooth.

I had some range issues initially, and I was not too keen on sleeping on top of my laptop :) In case you run into the same trouble, I had luck with that adapter.

> but how did you make the watch continue to send these measurements when the screen goes to sleep. are you using custom firmware?

Yes, as mentioned in https://github.com/InfiniTimeOrg/InfiniTime/discussions/1918#discussioncomment-7695789 and linked by @Schweini07, I used https://github.com/InfiniTimeOrg/InfiniTime/pull/1718 to get continuous background measurements.

**Reply by melvinredondotanis** [reply]
thank you, I will try to compile the version which is not yet merged but if I cannot do it I will try with the adapter

**Comment by melvinredondotanis** [comment]
I created a repo to gather all the information, measurements, codes etc.

https://github.com/MelvinRedondoTanis/SleepTracking-PineTime

I haven't put much into it yet

**Comment by melvinredondotanis** [comment]
In the end I didn't manage to compile. I could not yet do a test over a long period. for the moment I have not yet bought the adapter but in the meantime I tried to do a little function to detect a significant movement as simply as possible
https://github.com/MelvinRedondoTanis/SleepTracking-PineTime/blob/main/src/processing.py

**Comment by Schweini07** [comment]
Hi!
I've compiled PR #1718 for anybody that is interested. I am going to test and track the heart measurements for the next days now, while sleeping :smile:.
[pinetime-mcuboot-app-dfu-1.13.0.zip](https://github.com/InfiniTimeOrg/InfiniTime/files/13624924/pinetime-mcuboot-app-dfu-1.13.0.zip)

**Reply by melvinredondotanis** [reply]
Thank you so much

**Reply by tausen** [reply]
Did you manage to log some data, @Schweini07? If so, did it resemble those I plotted earlier? I'm curious whether the noisy heart rate measurements in my plot were due to operator error :)

**Reply by Schweini07** [reply]
Here is an example of a night about two weeks ago. I sadly forgot to document what kind of interval I was using for the heart tracking each night, but I assume this was with the 10s interval. Although I think Continuous and 10s don't have much of a difference.

This is the overall plot of the night, where one thing is quite obvious. The data only sometimes seems to be taken in a regular interval, other times there may be a difference of even an hour, before the next value is plotted.
![grafik](https://github.com/InfiniTimeOrg/InfiniTime/assets/47503977/e8dff44c-26e7-4c86-8d7c-571bf5dca683)

Here, we can see this in more detail. I am quite unsure to be honest, what might be the reason of this, especially because I feel the Gadgetbridge graph is far more accurate when using heart tracking during the day.
![grafik](https://github.com/InfiniTimeOrg/InfiniTime/assets/47503977/4b702a33-be26-4a83-84d5-bd1c27baadff)

**Reply by tausen** [reply]
Thanks, interesting. This is logged using gadgetbridge?
Could it be a bluetooth connection issue?

**Reply by Schweini07** [reply]
Yes, it is logged through Gadgetbridge. It would be pretty weird though if the problem occurs because of a bluetooth connection issue, as my phone was always laying just half a meter next to the watch.

**Comment by melvinredondotanis** [comment]
I ultimately think that the best thing is to work on an application dedicated to pinetime, even if simplified, sleep monitoring is too demanding for the power of pinetime

**Reply by Schweini07** [reply]
I'd oppose that. Certainly, the PineTime's battery will be drained immensely during a night, but with optimizations we can minimize that.
What I found out while testing the heart tracking is, that the battery drain for continuous tracking is about ~60% per night if I remember correctly. But if it's set to a 10 second interval it was about ~20-30%. I am sure we could set this interval to even 1 minute, and we would still get good results, whilst being easy on the battery.

The more prevalent problem I encountered during testing were inconsistent readings. For one hour during the night, the tracking was fine, but then for the next two hours it stayed at one value or shot up to a BPM of 150. I am unsure on how we could tackle this problem 😅.

**Reply by melvinredondotanis** [reply]
actually the battery can last longer set to 1 min but currently the sleep tracking on gadgetbridge works in the form of a signal like a sensor. to carry out sleep monitoring, the watch must therefore calculate the quality of sleep itself according to the data it collects and the very basic calculations must be joined to a more complex algorithm responsible for differentiating possible sensor bugs (quite common : I always have 1 to 2 hours per night) that's why I think a dedicated application is easier to set up. the application will relieve the watch.

**Reply by Schweini07** [reply]
Though if we'd iron out these sensor bugs, we should be able to provide Gadgetbridge, or any other companion app, with the right data for sleep tracking, no? And I'd think the calculations shouldn't be much of a drain factor for the watch.

With the sensor bug you mean that for 1-2 hours a night Gadgetbridge only shows the same Value, right? Because that is the problem I am having at least.

**Reply by tausen** [reply]
I think a PC application that takes PineTime sensor data in and generates sleep quality estimates out would be a good place to _start_. This would let us test the impact of sensor update interval on the sleep quality estimate and get a feel for the necessary computation and storage. It could also serve as a platform for choosing and tweaking sleep tracking algorithms. If at all possible, though, I agree with @Schweini07 that we should strive for a solution that can run on the PineTime.

> The more prevalent problem I encountered during testing were inconsistent readings. For one hour during the night, the tracking was fine, but then for the next two hours it stayed at one value or shot up to a BPM of 150.

This sounds like something we should dig into. Are there any open issues mentioning this? Is it possible the watch did not have good skin contact during the periods with poor readings? I did not see anything similar during my tests.

**Reply by melvinredondotanis** [reply]
making a mini app to develop algorithms is a good solution for the moment. how could we communicate, a discord or something else?

**Reply by Schweini07** [reply]
A platform for easier communication would be quite nice. Discord should be fine, but if there is a better alternative that would be great too. I always feel like on Discord information gets so easily lost :sweat_smile:

**Reply by melvinredondotanis** [reply]
I haven't found anything else, in the meantime let's use discord

**Comment by melvinredondotanis** [comment]
https://discord.com/invite/d6jvmXuh
